src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpInterop.cs

```diff
@@ -0,0 +1,290 @@
+                    break;
+                case "HEAD":
+                    method = Method.head();
+                    break;
+                case "POST":
+                    method = Method.post();
+                    break;
+                case "PUT":
+                    method = Method.put();
+                    break;
+                case "DELETE":
+                    method = Method.delete();
+                    break;
+                case "CONNECT":
+                    method = Method.connect();
+                    break;
+                case "OPTIONS":
+                    method = Method.options();
+                    break;
+                case "TRACE":
+                    method = Method.trace();
+                    break;
+                case "PATCH":
+                    method = Method.patch();
+                    break;
+                default:
+                    method = Method.other(requestMethod.Method);
+                    break;
+            }
+            return method;
+        }
+
+        public static Scheme ConvertScheme(Uri uri)
+        {
+            Scheme scheme;
+            switch (uri.Scheme)
+            {
+                case "":
+                case "http":
+                    scheme = Scheme.http();
+                    break;
+                case "https":
+                    scheme = Scheme.https();
+                    break;
+                default:
+                    scheme = Scheme.other(uri.Scheme);
+                    break;
+            }
+            return scheme;
+        }
+
+        public static string ConvertAuthority(Uri uri)
+        {
+            // `wasi:http/outgoing-handler` requires a non-empty authority,
+            // so we set one here:
+            if (string.IsNullOrEmpty(uri.Authority))
+            {
+                if (uri.Scheme == "https")
+                {
+                    return ":443";
+                }
+                else
+                {
+                    return ":80";
+                }
+            }
+            else
+            {
+                return uri.Authority;
+            }
+        }
+
+        public static Fields ConvertRequestHeaders(HttpRequestMessage request)
+        {
+            var headers = new List<(string, byte[])>();
+            foreach (var pair in request.Headers)
+            {
+                foreach (var value in pair.Value)
+                {
+                    headers.Add((pair.Key, Encoding.UTF8.GetBytes(value)));
+                }
+            }
+            if (request.Content is not null)
+            {
+                foreach (var pair in request.Content.Headers)
+                {
+                    foreach (var value in pair.Value)
+                    {
+                        headers.Add((pair.Key, Encoding.UTF8.GetBytes(value)));
+                    }
+                }
+            }
+            return Fields.FromList(headers);
```

Comment:
Consider wrapping this in a try/catch [like this](https://github.com/dotnet/runtimelab/blob/9d9ed59f2ec05b9430bdcb314e95adea0dc5ab16/src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpHandler.cs#L225-L251) in order to give the developer more information if the host rejects one or more headers.  I've been trying to improve exception messages bit-by-bit as I encounter them.

Reply:
Will do.
I will not dump the headers because they could be security tokens.